File: xkcd/src/main/java/xyz/jienan/xkcd/list/XkcdListGridAdapter.java
package xyz.jienan.xkcd.list;

import android.annotation.SuppressLint;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.percent.PercentFrameLayout;
import android.support.percent.PercentLayoutHelper;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.Priority;
import com.bumptech.glide.RequestManager;
import com.bumptech.glide.load.engine.DiskCacheStrategy;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import timber.log.Timber;
import xyz.jienan.xkcd.R;
import xyz.jienan.xkcd.model.XkcdModel;
import xyz.jienan.xkcd.model.XkcdPic;

public class XkcdListGridAdapter extends RecyclerView.Adapter<XkcdListGridAdapter.XkcdViewHolder> implements RecyclerViewFastScroller.BubbleTextGetter {

    private Context mContext;
    private RequestManager glide;
    private List<XkcdPic> pics = new ArrayList<>();
    private XkcdModel xkcdModel = XkcdModel.getInstance();

    public XkcdListGridAdapter(Context context) {
        mContext = context;
        glide = Glide.with(context);
    }

    public RequestManager getGlide() {
        return glide;
    }

    @NonNull
    @Override
    public XkcdViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(mContext).inflate(R.layout.item_xkcd_list, parent, false);
        return new XkcdViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull XkcdViewHolder holder, int position) {
        XkcdPic pic = pics.get(holder.getAdapterPosition());
        holder.bind(pic);
    }

    @Override
    public int getItemCount() {
        return pics == null ? 0 : pics.size();
    }

    public List<XkcdPic> getPics() {
        return pics;
    }

    @Override
    public String getTextToShowInBubble(int pos) {
        if (pics != null && pics.get(pos) != null && pics.get(pos).num > 0) {
            return String.valueOf(pics.get(pos).num);
        } else {
            return String.valueOf(pos + 1);
        }
    }

    public void updateData(List<XkcdPic> pics) {
        this.pics = pics;
        notifyDataSetChanged();
    }

    public XkcdPic getPic(int position) {
        return pics.get(position);
    }

    class XkcdViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.iv_item_xkcd_list)
        ImageView itemXkcdImageView;
        @BindView(R.id.tv_item_xkcd_num)
        TextView itemXkcdImageNum;

        XkcdViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        @SuppressLint("CheckResult")
        void bind(final XkcdPic pic) {
            PercentFrameLayout.LayoutParams layoutParams =
                    (PercentFrameLayout.LayoutParams) itemXkcdImageView.getLayoutParams();
            PercentLayoutHelper.PercentLayoutInfo info = layoutParams.getPercentLayoutInfo();
            int width = pic.width;
            int height = pic.height;

            info.aspectRatio = ((float) width) / height;
            layoutParams.height = 0;
            glide.load(pic.getTargetImg())
                    .asBitmap()
                    .diskCacheStrategy(DiskCacheStrategy.SOURCE)
                    .priority(Priority.HIGH)
                    .fitCenter()
                    .dontTransform()
                    .into(itemXkcdImageView);
            itemXkcdImageNum.setText(String.valueOf(pic.num));
            if (pic.isFavorite) {
                itemXkcdImageNum.setBackground(mContext.getResources().getDrawable(R.drawable.ic_heart_on));
            } else {
                itemXkcdImageNum.setBackground(mContext.getResources().getDrawable(R.drawable.item_num_bg));
            }
            if (width == 0 || height == 0) {
                xkcdModel.loadXkcd(pic.num)
                        .subscribe(xkcdPic -> {
                            if (xkcdPic != null) {
                                xkcdModel.updateSize(xkcdPic.num, xkcdPic.width, xkcdPic.height);
                            }
                        }, e -> Timber.e(e, "reload size error"));
            }
        }
    }
}

File: xkcd/src/main/java/xyz/jienan/xkcd/list/XkcdListGridAdapter.kt
package xyz.jienan.xkcd.list

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.percentlayout.widget.PercentFrameLayout
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.RequestManager
import kotlinx.android.synthetic.main.item_xkcd_list.view.*
import timber.log.Timber
import xyz.jienan.xkcd.R
import xyz.jienan.xkcd.base.glide.XkcdGlideUtils
import xyz.jienan.xkcd.model.XkcdModel
import xyz.jienan.xkcd.model.XkcdPic

internal class XkcdListGridAdapter : ListBaseAdapter<XkcdListGridAdapter.XkcdViewHolder>(), RecyclerViewFastScroller.BubbleTextGetter {

    private lateinit var glide: RequestManager

    var pics: List<XkcdPic>? = listOf()
        private set

    override var pauseLoading: Boolean
        set(pauseLoading) {
            if (pauseLoading) glide.pauseRequests() else glide.resumeRequests()
        }
        get() = glide.isPaused

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): XkcdViewHolder {
        glide = Glide.with(parent.context)
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_xkcd_list, parent, false)
        return XkcdViewHolder(view)
    }

    override fun onBindViewHolder(holder: XkcdViewHolder, position: Int) {
        holder.bind(pics!![holder.bindingAdapterPosition])
    }

    override fun getItemCount(): Int = pics?.size ?: 0

    override fun getTextToShowInBubble(pos: Int): String {
        return if (pics != null && pics!![pos].num > 0) {
            pics!![pos].num.toString()
        } else {
            (pos + 1).toString()
        }
    }

    fun updateData(pics: List<XkcdPic>) {
        this.pics = pics
        notifyDataSetChanged()
    }

    fun getPic(position: Int): XkcdPic? = pics?.get(position)

    internal inner class XkcdViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        @SuppressLint("CheckResult")
        fun bind(pic: XkcdPic) {
            val layoutParams = itemView.itemXkcdImageView!!.layoutParams as PercentFrameLayout.LayoutParams
            val info = layoutParams.percentLayoutInfo
            val width = pic.width
            val height = pic.height

            info.aspectRatio = width.toFloat() / height
            layoutParams.height = 0
            XkcdGlideUtils.load(glide, pic, pic.targetImg, itemView.itemXkcdImageView)
            itemView.itemXkcdImageNum?.text = pic.num.toString()
            itemView.itemXkcdImageNum?.background = if (pic.isFavorite) {
                ContextCompat.getDrawable(itemView.context, R.drawable.ic_heart_on)
            } else {
                ContextCompat.getDrawable(itemView.context, R.drawable.item_num_bg)
            }

            if (width == 0 || height == 0) {
                XkcdModel.loadXkcd(pic.num)
                        .subscribe({ xkcdPic ->
                            if (xkcdPic != null) {
                                XkcdModel.updateSize(xkcdPic.num, xkcdPic.width, xkcdPic.height)
                            }
                        }, { e -> Timber.e(e, "reload size error") })
            }
        }
    }
}